### ISSUE TITLE
`unnecessary_lambdas` constructor tear-off false postitives 

### ISSUE THREAD

**Issue by MohiuddinM** [dart-lang/linter#4914]
```
class A {
  A(int i);
}

void main() {
  // these are ok (but dart lint encourages tear-off)
  [].map((e) => A(e));
  [].map((e) => A.new(e));
  
  // this is an error: The argument type 'A Function(int)' can't be assigned to the parameter type 'dynamic Function(dynamic)'.
  [].map(A.new);
}
```

**Comment by eernstg** [comment]
> we should fix these false positives in the linter.

Interestingly, we have several design forces pushing each other around here.

`unnecessary_lambdas` might detect that the eta expansion has an effect because (as @lrhn mentioned) the bare `A.new` and `(x) => A.new(x)` have different signatures. So it's not unnecessary, and shouldn't be reported.

However, some other lint or option (perhaps `avoid_dynamic_calls`) should then be able to report that the invocation of `A.new(x)` in the body of the function literal may incur a run-time type error.

That should help nudging developers in the direction of the more concise and statically typed variant `A.new`.

Still, `[]` would be inferred to mean `<dynamic>[]` because it's a receiver (and hence there's no context type). So `A.new` will still give rise to a compile-time error. This may not be very helpful, but it will at least give the developer a heads-up that something about the typing needs to be refined.

For the future, if and when we get https://github.com/dart-lang/language/issues/3527, we will also be able to infer `[]` as `<int>[]`.

**Comment by srawlins** [comment]
Fixed with https://github.com/dart-lang/sdk/commit/293469ee22ba8cb62679310cdfb682019ffff5bd
